### ISSUE TITLE
filer.sync corrupted file in destination

### ISSUE THREAD

**Issue by michagruner** [seaweedfs/seaweedfs#3774]
**Describe the bug**
After doing
`./weed filer.sync -a=weed1:8888 -b=debianbackup:8888 -isActivePassive -a.filerProxy -b.filerProxy -concurrency=1`
every other file in my debianbackup-destination is corrupted. So for example when I try to download the file from the destination filer I get an unexpected EOF message from my browser. After renaming the file in the source-filer it gets synced correctly and I can see the correct content in the destination filer.

**System Setup**
debianbackup:

`Linux debianbackup 5.10.0-17-amd64 #1 SMP Debian 5.10.136-1 (2022-08-13) x86_64 GNU/Linux`

docker-compose.yaml:

```
mgruner@debianbackup:~/docker$ cat docker-compose.yml 
version: '2'

services:
  db:
    image: postgres:14-bullseye
    restart: unless-stopped
    ports:
      - 5432:5432
    environment:
      POSTGRES_PASSWORD: seaweed.Pwd
      POSTGRES_USER: seaweed
      POSTGRES_DB: seaweedfs
    volumes: 
      - /home/mgruner/postgres:/var/lib/postgresql/data

  master0:
    image: chrislusf/seaweedfs
    ports:
      - 9333:9333
      - 19333:19333
    volumes:
      - /home/mgruner/master:/data
    command: "master -ip=debianbackup -ip.bind=0.0.0.0 -mdir=/data/m1"
    restart: unless-stopped
  volume1:
    image: chrislusf/seaweedfs
    ports:
      - 8080:8080
      - 18080:18080
    volumes:
      - /home/mgruner/hetzner/volume1:/data
      #- /mnt/hetzner/volume1:/data
    command: 'volume -mserver="debianbackup:9333" -port=8080 -ip.bind=0.0.0.0 -ip=debianbackup -dir=/data/'
    restart: unless-stopped
    depends_on:
      - master0
  volume2:
    image: chrislusf/seaweedfs
    ports:
      - 8081:8081
      - 18081:18081
    volumes:
      - /home/mgruner/hetzner/volume2:/data
        #- /mnt/hetzner/volume2:/data
    command: 'volume -mserver="debianbackup:9333" -port=8081 -ip.bind=0.0.0.0 -ip=debianbackup -dir=/data/'
    restart: unless-stopped
    depends_on:
      - master0
  filer:
    image: chrislusf/seaweedfs
    ports:
      - 8888:8888
      - 18888:18888
    volumes:
      - /home/mgruner/filer:/data
      - ./filer.toml:/etc/seaweedfs/filer.toml
    command: 'filer -master="debianbackup:9333" -ip.bind=0.0.0.0 -ip=debianbackup -encryptVolumeData'
    restart: unless-stopped
    depends_on:
      - master0
      - db

```
filer.toml:

```
# A sample TOML config file for SeaweedFS filer store
# Used with "weed filer" or "weed server -filer"
# Put this file to one of the location, with descending priority
#    ./filer.toml
#    $HOME/.seaweedfs/filer.toml
#    /etc/seaweedfs/filer.toml

####################################################
# Customizable filer server options
####################################################
[filer.options]
# with http DELETE, by default the filer would check whether a folder is empty.
# recursive_delete will delete all sub folders and files, similar to "rm -Rf"
recursive_delete = false
# directories under this folder will be automatically creating a separate bucket
buckets_folder = "/buckets"

####################################################
# The following are filer store options
####################################################

[postgres2]
enabled = true
createTable = """
  CREATE TABLE IF NOT EXISTS "%s" (
    dirhash   BIGINT,
    name      VARCHAR(65535),
    directory VARCHAR(65535),
    meta      bytea,
    PRIMARY KEY (dirhash, name)
  );
"""
hostname = "debianbackup"
port = 5432
username = "seaweed"
password = "seaweed.Pwd"
database = "seaweedfs"          # create or use an existing database
schema = ""
sslmode = "disable"
connection_max_idle = 100
connection_max_open = 100
connection_max_lifetime_seconds = 0
# if insert/upsert failing, you can disable upsert or update query syntax to match your RDBMS syntax:
enableUpsert = true
upsertQuery = """INSERT INTO "%[1]s" (dirhash,name,directory,meta) VALUES($1,$2,$3,$4) ON CONFLICT (dirhash,name) DO UPDATE SET meta = EXCLUDED.meta WHERE "%[1]s".meta != EXCLUDED.meta"""

```
weed version 3.29

**Expected behavior**
Files at destination having the same content as files at source.

**Additional context**
attached the meta data of one specific file:

source-filer before renaming:
```
> meta.cat 023.JPG
{
  "name": "023.JPG",
  "isDirectory": false,
  "chunks": [
    {
      "fileId": "19,0b870a7570db6d",
      "offset": "0",
      "size": "3101015",
      "mtime": "1664221557843037646",
      "eTag": "DX0WUfldzPcHieGR2vxwRQ==",
      "sourceFileId": "",
      "fid": {
        "volumeId": 19,
        "fileKey": "755466",
        "cookie": 1970330477
      },
      "sourceFid": null,
      "cipherKey": "",
      "isCompressed": false,
      "isChunkManifest": false
    }
  ],
  "attributes": {
    "fileSize": "3101015",
    "mtime": "1664221557",
    "fileMode": 504,
    "uid": 232062,
    "gid": 232062,
    "crtime": "1664221557",
    "mime": "image/jpeg",
    "ttlSec": 0,
    "userName": "",
    "groupName": [],
    "symlinkTarget": "",
    "md5": "",
    "rdev": 0,
    "inode": "0"
  },
  "extended": {},
  "hardLinkId": "",
  "hardLinkCounter": 0,
  "content": "",
  "remoteEntry": null,
  "quota": "0"
}chunks 1 meta size: 126 gzip:154

```
destination filer before renaming:

```
> meta.cat 023.JPG
{
  "name": "023.JPG",
  "isDirectory": false,
  "chunks": [
    {
      "fileId": "26,09010f34af3440",
      "offset": "0",
      "size": "3101015",
      "mtime": "1664221557843037646",
      "eTag": "DX0WUfldzPcHieGR2vxwRQ==",
      "sourceFileId": "19,0b870a7570db6d",
      "fid": {
        "volumeId": 26,
        "fileKey": "590095",
        "cookie": 883897408
      },
      "sourceFid": {
        "volumeId": 19,
        "fileKey": "755466",
        "cookie": 1970330477
      },
      "cipherKey": "",
      "isCompressed": false,
      "isChunkManifest": false
    }
  ],
  "attributes": {
    "fileSize": "3101015",
    "mtime": "1664221557",
    "fileMode": 504,
    "uid": 232062,
    "gid": 232062,
    "crtime": "1664221557",
    "mime": "image/jpeg",
    "ttlSec": 0,
    "userName": "",
    "groupName": [],
    "symlinkTarget": "",
    "md5": "",
    "rdev": 0,
    "inode": "0"
  },
  "extended": {},
  "hardLinkId": "",
  "hardLinkCounter": 0,
  "content": "",
  "remoteEntry": null,
  "quota": "0"
}chunks 1 meta size: 158 gzip:186

```

source filer after renaming:

```
> meta.cat 023_new.JPG
{
  "name": "023_new.JPG",
  "isDirectory": false,
  "chunks": [
    {
      "fileId": "19,0b870a7570db6d",
      "offset": "0",
      "size": "3101015",
      "mtime": "1664221557843037646",
      "eTag": "DX0WUfldzPcHieGR2vxwRQ==",
      "sourceFileId": "",
      "fid": {
        "volumeId": 19,
        "fileKey": "755466",
        "cookie": 1970330477
      },
      "sourceFid": null,
      "cipherKey": "",
      "isCompressed": false,
      "isChunkManifest": false
    }
  ],
  "attributes": {
    "fileSize": "3101015",
    "mtime": "1664221557",
    "fileMode": 504,
    "uid": 232062,
    "gid": 232062,
    "crtime": "1664221557",
    "mime": "image/jpeg",
    "ttlSec": 0,
    "userName": "",
    "groupName": [],
    "symlinkTarget": "",
    "md5": "",
    "rdev": 0,
    "inode": "0"
  },
  "extended": {},
  "hardLinkId": "",
  "hardLinkCounter": 0,
  "content": "",
  "remoteEntry": null,
  "quota": "0"
}chunks 1 meta size: 130 gzip:158

```

destination filer after renaming:

```
> meta.cat 023_new.JPG
{
  "name": "023_new.JPG",
  "isDirectory": false,
  "chunks": [
    {
      "fileId": "170,11e4ff353fcb4a",
      "offset": "0",
      "size": "3101015",
      "mtime": "1664221557843037646",
      "eTag": "DX0WUfldzPcHieGR2vxwRQ==",
      "sourceFileId": "19,0b870a7570db6d",
      "fid": {
        "volumeId": 170,
        "fileKey": "1172735",
        "cookie": 893373258
      },
      "sourceFid": {
        "volumeId": 19,
        "fileKey": "755466",
        "cookie": 1970330477
      },
      "cipherKey": "",
      "isCompressed": false,
      "isChunkManifest": false
    }
  ],
  "attributes": {
    "fileSize": "3101015",
    "mtime": "1664221557",
    "fileMode": 504,
    "uid": 232062,
    "gid": 232062,
    "crtime": "1664221557",
    "mime": "image/jpeg",
    "ttlSec": 0,
    "userName": "",
    "groupName": [],
    "symlinkTarget": "",
    "md5": "",
    "rdev": 0,
    "inode": "0"
  },
  "extended": {},
  "hardLinkId": "",
  "hardLinkCounter": 0,
  "content": "",
  "remoteEntry": null,
  "quota": "0"
}chunks 1 meta size: 164 gzip:192

```

**Comment by chrislusf** [comment]
can you try skipping `-encryptVolumeData` option?

**Comment by michagruner** [comment]
I could for test purposes if that helps. The problem is: the data directory of the two volumes is located on a smb-mount from an untrusted samba-server. Therefore I'd like to have the volume data encrypted.

**Comment by michagruner** [comment]
Ok, did the setup without encryption. I think the problem might be somewhere in my source then, because the result is the same. On the sync I get a 

`1002 13:29:34.751339 filer_sink.go:125 replicate entry chunks /nextcloud/cgruner/files/Backup/20170604_170403.jpg: copy 15,0b90a0e5f3d1b2: upload data: read input: unexpected EOF
`
When I try to open the file in the destination it is corrupt obviously. When I open it in the source filer it works well. How can I dig deeper into the nature of the problem?

**Comment by michagruner** [comment]
My source cluster (3 different physical machines weed1/2/3 where 1 hosts the master and the filer, the other two host volumes looks like this:

docker-compose.yaml (of the master/filer node):
```

version: '2'

services:
  master0:
    image: chrislusf/seaweedfs
    ports:
      - 9333:9333
      - 19333:19333
    volumes:
      - /data/seaweedfs/master:/data
    command: "master -ip=weed1 -ip.bind=0.0.0.0 -port=9333 -peers=weed1:9333 -mdir=/data/m1 -defaultReplication=010 -volumeSizeLimitMB=15000"
    restart: unless-stopped
  volume1:
    image: chrislusf/seaweedfs
    ports:
      - 8080:8080
      - 18080:18080
    volumes:
      - /data/seaweedfs/volume/data1:/data
    command: 'volume -mserver="weed1:9333" -port=8080 -ip=weed1 -ip.bind=0.0.0.0 -dir=/data/ -publicUrl=weed1:8080 -dataCenter=Mylocation -rack=Proxmox max=0'
    restart: unless-stopped
    depends_on:
      - master0
  volume2:
    image: chrislusf/seaweedfs
    ports:
      - 8081:8081
      - 18081:18081
    volumes:
      - /data/seaweedfs/volume/data2:/data
    command: 'volume -mserver="weed1:9333" -port=8081 -ip=weed1 -ip.bind=0.0.0.0 -dir=/data/ -publicUrl=weed1:8081 -dataCenter=Mylocation -rack=Proxmox max=0'
    restart: unless-stopped
    depends_on:
      - master0
  filer:
#    image: chrislusf/seaweedfs:3.25_large_disk_rocksdb
    image: chrislusf/seaweedfs
    ports:
      - 8888:8888
      - 18888:18888
    volumes:
      - /data/seaweedfs/filer:/data
      - ./filer.toml:/etc/seaweedfs/filer.toml
    command: 'filer -master="weed1:9333" -ip=weed1 -ip.bind=0.0.0.0'
    restart: unless-stopped
    depends_on:
      - master0
```

filer.toml:

```

# A sample TOML config file for SeaweedFS filer store
# Used with "weed filer" or "weed server -filer"
# Put this file to one of the location, with descending priority
#    ./filer.toml
#    $HOME/.seaweedfs/filer.toml
#    /etc/seaweedfs/filer.toml

####################################################
# Customizable filer server options
####################################################
[filer.options]
# with http DELETE, by default the filer would check whether a folder is empty.
# recursive_delete will delete all sub folders and files, similar to "rm -Rf"
recursive_delete = false
# directories under this folder will be automatically creating a separate bucket
buckets_folder = "/buckets"

####################################################
# The following are filer store options
####################################################

[postgres2]
enabled = true
createTable = """
  CREATE TABLE IF NOT EXISTS "%s" (
    dirhash   BIGINT,
    name      VARCHAR(65535),
    directory VARCHAR(65535),
    meta      bytea,
    PRIMARY KEY (dirhash, name)
  );
"""
hostname = "dockerintern-p"
port = 5432
username = "postgres"
password = "seaweed.Pwd"
database = "seaweedfs"          # create or use an existing database
schema = ""
sslmode = "disable"
connection_max_idle = 100
connection_max_open = 100
connection_max_lifetime_seconds = 0
# if insert/upsert failing, you can disable upsert or update query syntax to match your RDBMS syntax:
enableUpsert = true
upsertQuery = """INSERT INTO "%[1]s" (dirhash,name,directory,meta) VALUES($1,$2,$3,$4) ON CONFLICT (dirhash,name) DO UPDATE SET meta = EXCLUDED.meta WHERE "%[1]s".meta != EXCLUDED.meta"""

```

could it have to do with the replication I use at the source cluster?

**Comment by michagruner** [comment]
For those coming to that issue and wanting to check their filer for health of the files I made a little script to do so and to also repair corrupt files on the destination filer by renaming them on the source filer leading to another sync:


weedcheck.sh
```
#!/bin/bash

OPTIND=1
optrepair=0

check()
{
	if tail -n 1 "$1" 2>/dev/null 1>/dev/null 
        then
#		echo "file is fine"
		retval=0
	else
#		echo "file corrupt"
		retval=1
	fi
	return $retval
}

repair()
{
	corruptfile="$1"
	destdir=$(echo "$corruptfile"|cut -d/ -f1)
	sourcedir="$2"
	#echo "Destdir $destdir"
	sourcefilename=$(echo "$corruptfile" |sed -e "s/${destdir}/${sourcedir}/g")
	echo $sourcefilename
	sourcetempname=$(echo "$sourcefilename" | sed -e "s/\./_\./g")
	echo $sourcetempname
	mv "$sourcefilename" "$sourcetempname"
	mv "$sourcetempname" "$sourcefilename"
}


while getopts "r:h" opt; do
  case "$opt" in
    r)  optrepair=1;optsourcedir="$OPTARG"
      ;;
    h)  echo "usage: weedcheck.sh [-r <sourcedirectory>] <file>  
	    checks if a file is readable on a seaweedfiler, option -r repairs the file by renaming it in the mounted filer in <sourcedirectory>"
	exit 0
      ;;
  esac
done

shift $((OPTIND -1))

file=$1

if check "$file"
then 
	echo 'file '"$file"' is fine'
else
	echo 'file '"$file"' is corrupt'
	if [ "$optrepair" == 1 ]
	then
		repair "$file" "$optsourcedir"
	fi
fi


```
How to use it:

- Start your sync process.
- Create two directories (in my case debianbackup for the destination filer, weed1 for the source filer)
- Mount the filers to the directories accordingly.
- Run weedcheck to see if there are corrupt files:
 `find debianbackup/nextcloud/cgruner/files/SofortUpload/Camera/2017/08/* -type f -exec sh -c "./weedcheck.sh \"{}\"" \;` 
- Run weedcheck with the repair-option and the sourcedirectory to "repair" them by renaming them back and forth:
 `find debianbackup/nextcloud/cgruner/files/SofortUpload/Camera/2017/08/* -type f -exec sh -c "./weedcheck.sh -r weed1 \"{}\"" \;`

Hope that helps.

**Comment by chrislusf** [comment]
The issue is because the `weed filer.sync` can not "see" the source volume servers. You can use this:

```
weed filer.sync -h
Example: weed filer.sync -a=<oneFilerHost>:<oneFilerPort> -b=<otherFilerHost>:<otherFilerPort>
Default Usage:
  -a.filerProxy
    	read and write file chunks by filer A instead of volume servers
  -b.filerProxy
    	read and write file chunks by filer B instead of volume servers
```

**Comment by michagruner** [comment]
I'm not sure this is the reason, because I already used the filerProxy as you can see in my first post. After renaming the file it syncs fine with the very same command and the same settings. I think it probably has something to do with the memory error you fixed earlier this week, because since I have the machine at 4GB I didn't see another "EOF..." message. The tricky thing is: the file gets created on the target filer with the correct file size displayed and all. So one could have the impression the file is healthy while there are obviously some chunks missing (you can see that at the shell where you do the weed mount when you tail the last line of a file; this is how I check if the file is healthy).

**Comment by chrislusf** [comment]
Was the sync process moving some large files around that time?

**Comment by michagruner** [comment]
I have the impression larger files (in my case often mp4 videos) are effected more often. One didn't even go through and the sync stopped working (that was why I opened the other issue with the large file). But other than that it seems very random and also smaller files are effected (the smallest I found was only 59k). It's really interesting, there seem to be no issues when I do the renaming and then the syncing gets done. But as I said that could also be the case because I upped the memory of my sync-vm to get that large file through in the meantime.
